### PR TITLE
Consolidate versioning for GPU hardware

### PR DESCRIPTION
**Pull request by whchung**
Different methods are used in GPU common runtime, StreamExecutor, and XLA to
track GPU hardware versions. To support upcoming ROCm SteramExecutor and AMDGPU
XLA, it's now consolidated to DeviceVersion in StreamExecutor in this commit.

### CODE REVIEW

**Comment by whchung** [comment]
@jlebar This is the PR created to consolidate GPU hardware versioning after our discussion in #20757. Now that particular PR could be greatly simplified. However this particular PR affects quite a few places in TensorFlow and XLA.

File: tensorflow/stream_executor/device_description.h [hunk 1]
```diff
@@ -156,8 +200,15 @@ class DeviceDescription {
...
   // Returns the CUDA compute capability if we're running on the CUDA platform.
   // If a CUDA compute capability is not available, the major version will be
-  // zero, and the return value will be false.
-  bool cuda_compute_capability(int *major, int *minor) const;
+  // zero, and the return value of DeviceVersion::is_valid() will be false.
+  //
+  // Returns the AMDGCN ISA version if we're running on the ROCm platform.
+  // Minor version would always be zero.
+  // If an AMDGCN ISA version is not available, the major version will be
+  // zero, and the return value of DeviceVersion::is_valid() will be false.
```
Comment (jlebar) [inline tensorflow/stream_executor/device_description.h]: * s/would always be/is always/

* s/the return value of DeviceDescription::is_valid() will be false/DeviceDescription::is_valid() will return false/

But even better, perhaps rewrite this comment so that it's not redundant with the comment on `DeviceVersion`?

> Gets the "hardware version" of this device, where we know how.  If we don't know how, the returned DeviceVersion's is_valid() returns false.

**Review by jlebar — CHANGES_REQUESTED** [review]

File: tensorflow/stream_executor/stream_executor_pimpl.cc [hunk 2]
```diff
@@ -333,10 +335,11 @@ bool StreamExecutor::GetConvolveBackwardFilterAlgorithms(
   if (!dnn_support) {
     return false;
   }
-  int cc_major, cc_minor;
-  GetDeviceDescription().cuda_compute_capability(&cc_major, &cc_minor);
+  DeviceVersion device_version =
+      GetDeviceDescription().device_hardware_version();
   return dnn_support->GetConvolveBackwardFilterAlgorithms(
-      with_winograd_nonfused, cc_major, cc_minor, out_algorithms);
```
[the remark this answers is not in the document]
Comment (whchung) [reply tensorflow/stream_executor/stream_executor_pimpl.cc]: As it touches `dnn.h` and I'd like to save it for another patch.

**Comment by whchung** [comment]
@jlebar Updated the PR per review comments. Please be cautious reviewing the static member function `DeviceVersion::Parse` as it would be used by bazel-generated header file `cuda_config.h` and I'm not very sure making its return type be `StatusOr<DeviceVersion>` a good idea.

File: tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc [hunk 3]
```diff
@@ -86,25 +87,25 @@ static string GetLibdeviceFilename(const string& libdevice_dir_path,
...
   int libdevice_version = 20;
   auto it = m->find(compute_capability);
   if (it != m->end()) {
     libdevice_version = it->second;
   } else {
-    LOG(WARNING) << "Unknown compute capability (" << compute_capability.first
-                 << ", " << compute_capability.second << ") ."
+    LOG(WARNING) << "Unknown compute capability (" << compute_capability
+                 << ") ."
                  << "Defaulting to libdevice for compute_" << libdevice_version;
```
Comment (jlebar) [inline tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc]: Nit: Please reflow string.

File: tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc [hunk 4]
```diff
@@ -113,28 +114,28 @@ static string GetLibdeviceFilename(const string& libdevice_dir_path,
...
   int sm_version = 30;
   auto it = m->find(compute_capability);
   if (it != m->end()) {
     sm_version = it->second;
   } else {
-    LOG(WARNING) << "Unknown compute capability (" << compute_capability.first
-                 << ", " << compute_capability.second << ") ."
+    LOG(WARNING) << "Unknown compute capability (" << compute_capability
+                 << ") ."
                  << "Defaulting to telling LLVM that we're compiling for sm_"
```
Comment (jlebar) [inline tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc]: Please reflow string.

File: tensorflow/contrib/fused_conv/kernels/fused_conv2d_bias_activation_op.cc [hunk 5]
```diff
@@ -322,11 +322,12 @@ void LaunchFusedConv2DBiasActivationOp<GPUDevice, T, BiasType, ScaleType>::
...
-    int cc_major, cc_minor;
-    stream->parent()->GetDeviceDescription().cuda_compute_capability(&cc_major,
-                                                                     &cc_minor);
+    DeviceVersion device_version =
+        stream->parent()->GetDeviceDescription().device_hardware_version();
     OP_REQUIRES(
-        ctx, ((cc_major == 6 && cc_minor >= 1) || cc_major > 6),
+        ctx,
+        ((device_version.major_part == 6 && device_version.minor_part >= 1) ||
+         device_version.major_part > 6),
```
Comment (jlebar) [inline tensorflow/contrib/fused_conv/kernels/fused_conv2d_bias_activation_op.cc]: `device_version >= DeviceVersion(6, 1)`?

File: tensorflow/core/common_runtime/gpu/gpu_device.cc [hunk 6]
```diff
@@ -1394,16 +1361,11 @@ Status BaseGPUDeviceFactory::GetValidDeviceIds(
...
-      cc_major = 0;
-      cc_minor = 0;
-    }
+    DeviceVersion device_version = description.device_hardware_version();
     LOG(INFO) << "Found device " << i << " with properties: "
-              << "\nname: " << description.name() << " major: " << cc_major
-              << " minor: " << cc_minor
+              << "\nname: " << description.name()
+              << " major: " << device_version.major_part
+              << " minor: " << device_version.minor_part
```
Comment (jlebar) [inline tensorflow/core/common_runtime/gpu/gpu_device.cc]: Might as well use DeviceVersion::operator<< here.

File: tensorflow/stream_executor/stream_executor_pimpl.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (jlebar) [reply tensorflow/stream_executor/stream_executor_pimpl.cc]: sgtm

File: third_party/gpus/cuda_configure.bzl [hunk 7]
```diff
@@ -1015,7 +1015,7 @@ def _create_dummy_repository(repository_ctx):
             "%{cuda_version}": _DEFAULT_CUDA_VERSION,
             "%{cudnn_version}": _DEFAULT_CUDNN_VERSION,
             "%{cuda_compute_capabilities}": ",".join([
-                "CudaVersion(\"%s\")" % c
+                "DeviceVersion::Parse(\"%s\")" % c
```
Comment (jlebar) [inline third_party/gpus/cuda_configure.bzl]: We can restore the old behavior by making this `DeviceVersion::Parse(%s).ValueOrDie()`.  I admit I don't understand what this is doing or how to test it, though.  We'll need someone from the build team to look at this.

\summon @gunan

File: tensorflow/core/common_runtime/gpu/gpu_device.cc [hunk 8]
```diff
@@ -1316,7 +1283,7 @@ std::vector<CudaVersion> GetSupportedCudaComputeCapabilities() {
 #undef TF_XSTRING
   auto extra_capabilities = str_util::Split(extra_cuda_caps, ',');
   for (const auto& capability : extra_capabilities) {
-    cuda_caps.push_back(CudaVersion(capability));
+    cuda_caps.push_back(DeviceVersion::Parse(capability));
```
Comment (jlebar) [inline tensorflow/core/common_runtime/gpu/gpu_device.cc]: This shouldn't compile?  We need to "unwrap" the StatusOr one way or another.

File: tensorflow/stream_executor/device_description.h [hunk 9]
```diff
@@ -20,18 +20,72 @@ limitations under the License.
...
+    size_t dot_pos = version_name.find('.');
+    CHECK(dot_pos != string::npos)
+        << "Illegal version name: [" << version_name << "]";
+    port::StringPiece major_str = version_name.substr(0, dot_pos);
+    CHECK(port::safe_strto32(major_str.data(), &major_part))
+        << "Illegal version name: [" << version_name << "]";
+    port::StringPiece minor_str = version_name.substr(dot_pos + 1);
+    CHECK(port::safe_strto32(minor_str.data(), &minor_part))
+        << "Illegal version name: [" << version_name << "]";
+    return DeviceVersion(major_part, minor_part);
```
Comment (jlebar) [inline tensorflow/stream_executor/device_description.h]: The idea was to return an error rather than CHECK (which is a fatal runtime assertion).

You'd probably want to return FailedPreconditionError, see stream_executor/lib/status.h.

File: tensorflow/stream_executor/device_description.h [hunk 10]
```diff
@@ -154,10 +208,10 @@ class DeviceDescription {
   // Micro Devices, Inc.", or "GenuineIntel".
   const string &device_vendor() const { return device_vendor_; }
 
-  // Returns the CUDA compute capability if we're running on the CUDA platform.
-  // If a CUDA compute capability is not available, the major version will be
-  // zero, and the return value will be false.
-  bool cuda_compute_capability(int *major, int *minor) const;
+  // Returns the hardware version of the device.
```
Comment (jlebar) [inline tensorflow/stream_executor/device_description.h]: At this point we should just remove the comment.  That's fine, it's obvious what it does.

**Comment by whchung** [comment]
@jlebar revised the PR again per comments. I'm still somewhat uncomfortable with changes made in `cuda_configure.bzl` because for obvious reason I can't really test it.

File: tensorflow/core/common_runtime/gpu/gpu_device.cc [hunk 11]
```diff
@@ -1316,7 +1283,9 @@ std::vector<CudaVersion> GetSupportedCudaComputeCapabilities() {
 #undef TF_XSTRING
   auto extra_capabilities = str_util::Split(extra_cuda_caps, ',');
   for (const auto& capability : extra_capabilities) {
-    cuda_caps.push_back(CudaVersion(capability));
+    DeviceVersion device_version =
+        DeviceVersion::Parse(capability).ValueOrDie();
+    cuda_caps.push_back(device_version);
```
Comment (jlebar) [inline tensorflow/core/common_runtime/gpu/gpu_device.cc]: Nit, recommend inlining the `device_version` variable:

`cuda_caps.push_back(DeviceVersion::Parse(capability).ValueOrDie())`

File: tensorflow/stream_executor/cuda/cuda_blas.cc [hunk 12]
```diff
@@ -1861,12 +1861,11 @@ bool CUDABlas::DoBlasGemm(
 
   bool use_tensor_ops = false;
 #if CUDA_VERSION >= 9000
-  int cc_major, cc_minor;
-  stream->parent()->GetDeviceDescription().cuda_compute_capability(&cc_major,
-                                                                   &cc_minor);
+  DeviceVersion device_version;
+  stream->parent()->GetDeviceDescription().device_hardware_version();
```
Comment (jlebar) [inline tensorflow/stream_executor/cuda/cuda_blas.cc]: This does not look correct?  You mean to assign to the device_version variable, right?

File: tensorflow/stream_executor/cuda/cuda_dnn.cc [hunk 13]
```diff
@@ -2984,7 +2984,7 @@ bool CudnnSupport::DoFusedConvolve(
     const dnn::AlgorithmConfig& algorithm_config,
     dnn::ProfileResult* output_profile_result) {
   int cc_major, cc_minor;
-  stream->parent()->GetDeviceDescription().cuda_compute_capability(&cc_major,
+  stream->parent()->GetDeviceDescription().device_hardware_version(&cc_major,
```
Comment (jlebar) [inline tensorflow/stream_executor/cuda/cuda_dnn.cc]: This should be a build error?

**Review by jlebar — CHANGES_REQUESTED** [review]
> I'm still somewhat uncomfortable with changes made in cuda_configure.bzl because for obvious reason I can't really test it.

I also don't know how to test it, but maybe @yifeif can help.

**Comment by whchung** [comment]
@jlebar Thanks. Updated the PR once again per review comments.

**Comment by jlebar** [comment]
LGTM, but I'd like @yifeif or @gunan to sign off on the build system changes, so not approving.

Also given that we had build errors it would be nice if the tests would complete.  I don't know how to kick off another set of tests, though.  Or maybe that happens automatically?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @qlzh727: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Review by gunan — DISMISSED** [review]
As long as it passes presubmits, I am ok with the changes to bzl files.
We are deprecating cmake with this release anyway, so I am willing to just check tests for that too.

**Comment by qlzh727** [comment]
Please resolve the merge conflict before we kick off the tests and merge process.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by whchung** [comment]
@jlebar / @gunan after some unexpected hiatus I'm back working on this PR. I've rebased and fixed all known failures on CUDA path. I'm wondering could we give this PR another test on your end?

**Comment by jlebar** [comment]
>  jlebar added the kokoro:force-run  label just now

No idea if this does what I was hoping, but it's worth a try.  :)

**Comment by whchung** [comment]
@jlebar I just fixed errors in some test targets. Notice there is one failure which i don't think has anything to do with this PR:

```
//tensorflow/python/estimator:keras_test
```

also i noticed the syntax asked by `clang-format` is a bit different on CI now. could you let me know the latest syntax definition is?

**Comment by jlebar** [comment]
> also i noticed the syntax asked by clang-format is a bit different on CI now. could you let me know the latest syntax definition is?

/me summons @yifeif 

> could you help re-run the tests?

OK.  @yifeif is there any way to make this process of rerunning the tests less work for us?  Like, is it possible to give whchung the ability to kick tests himself?

**Comment by whchung** [comment]
@jlebar / @yifeif I fixed the errors found in the last round of tests. Wondering is it possible for me to apply `kokoro:force-run` label?

Also I'm looking forward to hearing from @yifeif wrt #20277.

**Comment by jlebar** [comment]
> Wondering is it possible for me to apply kokoro:force-run label?

Yeah, I know this is not a good experience for you as our partner.  :(  We're currently discussing internally whether and how it might be technically possible to allow this, within our security constraints.  I'll let you know.

**Comment by yifeif** [comment]
Sorry for the delay @whchung, I was on vacation last week. I replied in #20277. We are using clang-format  version 3.9 with -style=google in CI. The failure in `//tensorflow/python/estimator:keras_test` should be irrelevant.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @qlzh727: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by tensorflowbutler** [comment]
It has been 43 days that this pull-request has stalled. Please create a new pull-request with the requested changes.
